xla/service/gpu/transforms/windowed_einsum_handler.cc

```diff
@@ -492,6 +511,125 @@ absl::Status PostProcessUnrolledLoop(HloInstruction* loop, int64_t stream_id) {
       TF_RETURN_IF_ERROR(UpdateDotAndConsumerConfig(inst, stream_id));
       ++stream_id;
     }
+    // If dot's result is accumulated, this means we found a loop with
+    // contracting dim sharded.
+    HloInstruction* partial_dot;
+    if (Match(inst, m::AddAnyOrder(m::Op(),
+                                   m::Dot(&partial_dot, m::Op(), m::Op())))) {
+      partial_accumulations.push_back(partial_dot);
+    }
```

Comment:
Move all the added logic into a dedicated function, which turns many adds inside the unrolled while-loop to a reduce outside the unrolled while-loop. A possible name could be MoveReductionOutsideLoop.

Reply:
Sounds good, moved